### DISCUSSION TITLE
how does vllm handle wrong tokens in speculative decoding?

### DISCUSSION THREAD

**Discussion by Tomorrowdawn** [vllm-project/vllm#4284] [Q&A]
The core problem is, accepted tokens in different sentences of same batch are different.
 I tried to search a lot about batched speculative decoding and I found most inference frameworks ignore this problem by simply setting batch size == 1, except vllm. I read the code in vllm/spec_decode, and it seems vllm fill all wrong tokens with `-1`. I am still unclear about:

1. if there is other techniques to remove those -1s? 
2. I know a trivial method is to mask these -1s off, if vllm does the same, how does it affects the performance?

thanks.

**Comment by cadedaniel** [comment]
Before the engine appends token ids to sequences, it removes -1 tokens. The logic is here:
https://github.com/vllm-project/vllm/blob/34128a697ed2dd4d88a9829f35445fbfc5b85c64/vllm/engine/output_processor/multi_step.py#L68

**Reply by cadedaniel** [reply]
This allows speculative decoding to fit within continuous batching scheduling.

**Reply by Tomorrowdawn** [reply]
Thanks! I found this PR #3103 which introduces the "batch expansion" used for verification and I also reviewed the relevant code in spec_decode/batch_expansion.py and worker/worker. From the image you provided, it appears that batch expansion splits the output sentence into a group of sentences, but tokens "1", "2", and "3" are not present in the kvcache because the target model hasn't executed the forward pass on them yet. 
![](https://private-user-images.githubusercontent.com/950914/308838003-f24376e0-9a47-429b-9aae-0a8674a1750f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM5NzE0NzIsIm5iZiI6MTcxMzk3MTE3MiwicGF0aCI6Ii85NTA5MTQvMzA4ODM4MDAzLWYyNDM3NmUwLTlhNDctNDI5Yi05YWFlLTBhODY3NGExNzUwZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQyNFQxNTA2MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MjA4YTM1NGQ0Mjk1M2JjYTNiZjhjNmM3YzIwY2NhN2JkYzZjYTUzZTc4YWUyY2U1MDA2OGNjMmVmYjM3YWVjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-jlFt7hM0FBrZu1YtAry64p7-JfT3IChSYOzEBbQ9E0)
I was wondering if you could further explain the underlying mechanics of this process? I'm quite interested in gaining a deeper understanding of how it works. Any additional insights you could provide would be greatly appreciated.

**Reply by cadedaniel** [reply]
Your description is correct! The KV for tokens 1,2,3 is populated in the fwd pass of the target model, which occurs after batch expansion and before batch contraction.

For more details, check out this doc. 
It doesn't have any treatment of the draft model KV, but explains how the target model KV is populated using lookahead allocation. https://docs.google.com/document/d/1Z9TvqzzBPnh5WHcRwjvK2UEeFeq5zMZb5mFE8jR0HCs/edit#heading=h.1fjfb0donq5a

**Comment by Tomorrowdawn** [comment]
> Thanks! I found this PR #3103 which introduces the "batch expansion" used for verification and I also reviewed the relevant code in spec_decode/batch_expansion.py and worker/worker. From the image you provided, it appears that batch expansion splits the output sentence into a group of sentences, but tokens "1", "2", and "3" are not present in the kvcache because the target model hasn't executed the forward pass on them yet. ![](https://private-user-images.githubusercontent.com/950914/308838003-f24376e0-9a47-429b-9aae-0a8674a1750f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM5NzE0NzIsIm5iZiI6MTcxMzk3MTE3MiwicGF0aCI6Ii85NTA5MTQvMzA4ODM4MDAzLWYyNDM3NmUwLTlhNDctNDI5Yi05YWFlLTBhODY3NGExNzUwZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQyNFQxNTA2MTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MjA4YTM1NGQ0Mjk1M2JjYTNiZjhjNmM3YzIwY2NhN2JkYzZjYTUzZTc4YWUyY2U1MDA2OGNjMmVmYjM3YWVjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-jlFt7hM0FBrZu1YtAry64p7-JfT3IChSYOzEBbQ9E0) I was wondering if you could further explain the underlying mechanics of this process? I'm quite interested in gaining a deeper understanding of how it works. Any additional insights you could provide would be greatly appreciated.

[these lines](https://github.com/vllm-project/vllm/blob/468d761b32e3b3c5d64eeaa797e54ab809b7e50c/vllm/worker/model_runner.py#L449C1-L452C54) suggests only the last token id would be appended into "input_ids". For example, in the last sentence, I don't understand how target model compute the key & value for 1, 2, 3.

**Reply by cadedaniel** [reply]
In batch expansion we create a new vLLM sequence for each token to be scored. So a single sequence with spec tokens 1,2,3 becomes three sequences. This means that there is still one input id per sequence, namely, 1, 2, and 3.

Because QKV projection happens before saving to KV cache in each layer, which happens before each attention computation, we can safely do this without impacting the computation result.
